backend/api/Areas/Keycloak/Controllers/UserController.cs

```diff
@@ -1,96 +1,261 @@
     {
         #region Variables
         private readonly ILogger<UserController> _logger;
+        private readonly IPimsService _pimsService;
         private readonly IPimsAdminService _pimsAdminService;
         private readonly IMapper _mapper;
-        private readonly Pims.Api.Configuration.KeycloakOptions _optionsKeycloak;
-        private readonly IKeycloakRequestClient _requestClient;
+        private readonly IKeycloakAdmin _keycloakAdmin;
         #endregion
 
         #region Constructors
         /// <summary>
         /// Creates a new instance of a UserController class.
         /// </summary>
         /// <param name="logger"></param>
+        /// <param name="pimsService"></param>
         /// <param name="pimsAdminService"></param>
         /// <param name="mapper"></param>
-        /// <param name="optionsKeycloak"></param>
-        /// <param name="requestClient"></param>
-        public UserController(ILogger<UserController> logger, IPimsAdminService pimsAdminService, IMapper mapper, IOptionsMonitor<Pims.Api.Configuration.KeycloakOptions> optionsKeycloak, IKeycloakRequestClient requestClient)
+        /// <param name="keycloakAdmin"></param>
+        public UserController(ILogger<UserController> logger, IPimsService pimsService, IPimsAdminService pimsAdminService, IMapper mapper, IKeycloakAdmin keycloakAdmin)
         {
             _logger = logger;
+            _pimsService = pimsService;
             _pimsAdminService = pimsAdminService;
             _mapper = mapper;
-            _optionsKeycloak = optionsKeycloak.CurrentValue;
-            _requestClient = requestClient;
+            _keycloakAdmin = keycloakAdmin;
         }
         #endregion
 
         #region Endpoints
         /// <summary>
-        /// Fetch a list of users from keycloak.
+        /// Sync the user for the specified 'id' from keycloak with PIMS.
+        /// If the user does not exist in keycloak it will return a 400-BadRequest.
+        /// If the user does not exist in PIMS it will add it.
+        /// Also links the user to the appropriate groups it is a member of within keycloak.!--
+        /// If the group does not exist in PIMS it will add it.
         /// </summary>
+        /// <param name="id"></param>
+        /// <exception type="KeyNotFoundException">The user does not exist in keycloak.</exception>
+        /// <returns></returns>
+        [HttpPost("sync/{id}")]
+        public async Task<IActionResult> SyncUserAsync(Guid id)
+        {
+            var kuser = await _keycloakAdmin.GetUserAsync(id) ?? throw new KeyNotFoundException();
+            var kgroups = await _keycloakAdmin.GetUserGroupsAsync(id);
+
+            var euser = _pimsAdminService.User.Find(id);
+            if (euser == null)
+            {
+                // The user does not exist in PIMS, it needs to be added.
+                euser = _mapper.Map<Entity.User>(kuser);
+                euser.DisplayName = $"{kuser.LastName}, {kuser.FirstName}";
```

Comment:
normally I'd say we should put these in automapper, but I imagine you aren't interested in hearing that right now.